Add staking accounts to the membership pallet.

shamil-gadelshin opened the pull request:
### Changes
- add `confirm_staking_account` extrinsic
- add `add_staking_account_candidate` extrinsic
- add `remove_staking_account` extrinsic
- add `StakingAccountValidator` trait to the common module
- implement `StakingAccountValidator` in the membership module
- refactor staking handler and working group

[Main issue](https://github.com/Joystream/joystream/issues/1736)

shamil-gadelshin committed:
runtime: membership: Add staking_account_ids map.

shamil-gadelshin committed:
runtime: membership: Add add_staking_account_candidate extrinsic.

shamil-gadelshin committed:
runtime: membership: Add tests for add_staking_account_candidate extrinsic.

shamil-gadelshin committed:
runtime: membership: Add remove_staking_account extrinsic.

shamil-gadelshin committed:
runtime: membership: Add confirm_staking_account extrinsic.

shamil-gadelshin committed:
runtime: Add staking account validation for a member.

shamil-gadelshin committed:
runtime: staking-handler: Remove obsolete dependency.

shamil-gadelshin committed:
runtime: membership: Add tests.

- add tests for the is_member_staking_account_works() method.

bedeho commented on runtime-modules/membership/src/lib.rs:
```diff
@@ -684,6 +741,88 @@ decl_module! {
+            let membership = Self::ensure_membership(member_id)?;
+
+            ensure!(
+                membership.staking_account_exists(&staking_account_id),
```
Should we not check that its not confirmed already?

shamil-gadelshin replied:
Ok. I added this check.

bedeho reviewed (commented):
Currently a single staking account can be associated with an unlimited number of memberships. While in principle it _may_ be that this still is safe, as long as all staking on such accounts uses `is_account_free_of_conflicting_stakes` implementation that rejects reuse a lock type with itself, but I am still not 100% sure about that, and also its not the mental model we want to encourage. Having a storage map from equivalent to `account => member_idXstatus` would solve the problem, and also remove need for any staking account cap.

What do you think?

shamil-gadelshin committed:
runtime: membership: Add check for double staking account confirmation.

shamil-gadelshin committed:
runtime: membership: Modify staking accounts.

- change staking account verification to the double map.

shamil-gadelshin commented:
I changed staking account verification to the double map. I removed the limit for the staking account number and added "staking account could be used only once in the system" check.

shamil-gadelshin committed:
runtime: membership: Change double map to map for staking accounts.

shamil-gadelshin committed:
runtime: membership: Change staking account workflow.

shamil-gadelshin commented:
Review fixes:
- changed double map to map for the staking account registration
- changed the workflow (adding and removing account requires staking account origin, confirmation - member origin).

bedeho reviewed (approved):
Excellent work.
